plugins/GSdx/res/tfx.fx

```diff
@@ -517,76 +518,61 @@ float4 sample_color(float2 st)
 		t = c[0];
 	}
 
-	return t;
+	return trunc(t * 255.0f + 0.05f);
 }
 
-float4 tfx(float4 t, float4 c)
+float4 tfx(float4 T, float4 C)
 {
-	if(PS_TFX == 0)
-	{
-		if(PS_TCC)
-		{
-			c = c * t * 255.0f / 128;
-		}
-		else
-		{
-			c.rgb = c.rgb * t.rgb * 255.0f / 128;
-		}
-	}
-	else if(PS_TFX == 1)
-	{
-		if(PS_TCC)
-		{
-			c = t;
-		}
-		else
-		{
-			c.rgb = t.rgb;
-		}
-	}
-	else if(PS_TFX == 2)
-	{
-		c.rgb = c.rgb * t.rgb * 255.0f / 128 + c.a;
+	float4 C_out;
+	float4 FxT = trunc(trunc(C) * T / 128.0f);
 
-		if(PS_TCC)
-		{
-			c.a += t.a;
-		}
-	}
-	else if(PS_TFX == 3)
-	{
-		c.rgb = c.rgb * t.rgb * 255.0f / 128 + c.a;
+#if (PS_TFX == 0)
+	C_out = FxT;
+#elif (PS_TFX == 1)
+	C_out = T;
+#elif (PS_TFX == 2)
+	C_out.rgb = FxT.rgb + C.a;
+	C_out.a = T.a + C.a;
+#elif (PS_TFX == 3)
+	C_out.rgb = FxT.rgb + C.a;
+	C_out.a = T.a;
+#else
+	C_out = C;
+#endif
 
-		if(PS_TCC)
-		{
-			c.a = t.a;
-		}
-	}
+#if (PS_TCC == 0)
+	C_out.a = C.a;
+#endif
 
-	return saturate(c);
+#if (PS_TFX == 0) || (PS_TFX == 2) || (PS_TFX == 3)
+	// Clamp only when it is useful
+	C_out = min(C_out, 255.0f);
+#endif
+
+	return C_out;
 }
 
-void atst(float4 c)
+void atst(float4 C)
 {
-	float a = trunc(c.a * 255 + 0.01);
+	float a = C.a;
 
 #if 0
-    switch(Uber_ATST) {
-        case 0:
-            break;
-        case 1:
-            if (a > AREF) discard;
-            break;
-        case 2:
-            if (a < AREF) discard;
-            break;
-        case 3:
-            if (abs(a - AREF) > 0.5f) discard;
-            break;
-        case 4:
-            if (abs(a - AREF) < 0.5f) discard;
-            break;
-    }
+	switch(Uber_ATST) {
+		case 0:
+			break;
+		case 1:
+			if (a > AREF) discard;
+			break;
+		case 2:
+			if (a < AREF) discard;
+			break;
+		case 3:
+			if (abs(a - AREF) > 0.5f) discard;
+			break;
+		case 4:
+			if (abs(a - AREF) < 0.5f) discard;
+			break;
+	}
```

Comment:
I'd prefer if this formatting change was moved to the first commit.

Otherwise the commit looks ok.

Reply:
I think it's done now. It required some git trickery I'm not comfortable with, so I hope no change got lost.